### PR TITLE
Fix memeory leak in MessageFramer

### PR DESCRIPTION
**Pull request by louiscryan**
Fixes

https://github.com/grpc/grpc-java/issues/336

### CODE REVIEW

**Comment by louiscryan** [comment]
@buchgr @ejona86 

We've actually always had a leak, it was just very slow.

**Comment by buchgr** [comment]
@louiscryan aahh you were faster. I just realized the same thing :-).

**Comment by ejona86** [comment]
Why aren't we freeing it? It seems that is the core issue.

**Comment by buchgr** [comment]
@louiscryan your fix looks good to me. Just realized that I introduced this leak myself :D.

@ejona86 because of this line https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/transport/AbstractServerStream.java#L145

**Comment by ejona86** [comment]
I'm not comfortable with change.

It seems we need an else then in AbstractServerStream that does `close()`. `allocator.buffer(0, 0)` does use an empty buffer in Netty, but it also still must be freed.

**Comment by buchgr** [comment]
@ejona86 oh yes you are right, totally forgot about this. yeah an else should do the trick.

File: netty/src/main/java/io/grpc/transport/netty/NettyWritableBufferAllocator.java [hunk 1]
```diff
@@ -54,9 +56,11 @@
   private static final int MAX_BUFFER = 1024 * 1024;
 
   private final ByteBufAllocator allocator;
+  private final NettyWritableBuffer emptyBuffer;
 
   NettyWritableBufferAllocator(ByteBufAllocator allocator) {
     this.allocator = allocator;
+    emptyBuffer = new NettyWritableBuffer(new EmptyByteBuf(allocator));
```
Comment (nmittler) [inline netty/src/main/java/io/grpc/transport/netty/NettyWritableBufferAllocator.java]: Could you just use `Unpooled.EMPTY_BUFFER`?

**Comment by nmittler** [comment]
@ejona86 I believe Unpooled.EMPTY_BUFFER should allow calls to release.

**Comment by ejona86** [comment]
I'm surprised that @buchgr experienced this, given we shouldn't be getting in that code path at all in the benchmarks. That zero-length flush is optimized away in [ChannelImpl](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/ChannelImpl.java#L311).

I thought maybe @louiscryan's framing change may increase the occurrence, but we don't flush when the frame is full. Instead we flush [when the frame is full when we have something to write](https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/transport/MessageFramer.java#L184).

**Comment by ejona86** [comment]
The Netty allocator is already optimized to return zero-buffers when requested for 0 initial and 0 max capacity buffers.

**Comment by ejona86** [comment]
https://github.com/netty/netty/blob/master/buffer/src/main/java/io/netty/buffer/AbstractByteBufAllocator.java#L152

**Comment by buchgr** [comment]
@ejona86 @louiscryan check this line https://github.com/grpc/grpc-java/blob/master/netty/src/main/java/io/grpc/transport/netty/NettyWritableBufferAllocator.java#L64

**Comment by ejona86** [comment]
@buchgr, yeah, I just now saw that. I was trying really hard to figure out if @louiscryan's framing change started the behavior. It seems like we should add a optimization to NettyWritableBufferAllocator instead of creating a new empty() method.

**Comment by buchgr** [comment]
@ejona86 yes I agree. add a if(capacityHint == 0) check or so.

**Comment by ejona86** [comment]
@buchgr, although I am still wondering why we are trying to allocate a zero-length buffer. That should only be necessary for streaming calls when we close the stream.

**Comment by buchgr** [comment]
@ejona86 here is the complete stack frame of how the allocation happend https://github.com/grpc/grpc-java/issues/336 ... it's because we called close in the unary call.

**Comment by buchgr** [comment]
@ejona86 @louiscryan so basically for every close we would now allocate a 4K buffer, which could help explaining the increased memory usage.

**Comment by louiscryan** [comment]
I guess we're both responsible for the issue, my change compounded Jakobs.

I have a test that easily reproduces the issue so I'll work on fixing the server stream and revert the 'empty' buffer magic

**Comment by louiscryan** [comment]
The more immediate stack for the issue is

  java.lang.Thread.State: RUNNABLE
	  at io.grpc.transport.MessageFramer.commitToSink(MessageFramer.java:240)
	  at io.grpc.transport.MessageFramer.close(MessageFramer.java:221)
	  at io.grpc.transport.AbstractStream.closeFramer(AbstractStream.java:136)
	  at io.grpc.transport.AbstractServerStream.close(AbstractServerStream.java:107)
	  at io.grpc.ServerImpl$ServerCallImpl.close(ServerImpl.java:434)
	  at io.grpc.benchmarks.netty.EndToEnd$1$1.onHalfClose(EndToEnd.java:164)
	  at io.grpc.ServerImpl$ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerImpl.java:480)
	  at io.grpc.ServerImpl$JumpToApplicationThreadServerStreamListener$2.run(ServerImpl.java:370)
	  at io.grpc.SerializingExecutor$TaskRunner.run(SerializingExecutor.java:152)
	  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1113)
	  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:588)
	  at java.lang.Thread.run(Thread.java:724)

**Comment by ejona86** [comment]
The bug is much older than either of your changes. It is actually from _October_ in commit fcab51cc. We've been insulated from it because Netty was returning the empty buffer. @louiscryan's changes just removed the empty buffer optimization which allowed it to surface.

**Comment by louiscryan** [comment]
I'm not a big fan of checking capacity == 0 as a special case in the allocator for what is really a special case in the MessageFramer so I've left the 'empty' methods given that they also save us an object allocation.

PTAL

**Comment by buchgr** [comment]
@louiscryan LGTM

**Comment by ejona86** [comment]
:( Even without empty() we don't need any extra allocation. We can return a singleton from allocate.

**Comment by louiscryan** [comment]
That's true but at least with empty the semantics are clearer and the enforcement stricter.

I'd still need the specialized subclass for OkHttp to enable a singleton as there are currently no capacity constraints on those buffers.

I'd be happy to get rid of empty() when the funky dependency causing trailers to be triggered the way they are today for the server is fixed. Suggestions in that direction welcome

**Comment by ejona86** [comment]
I don't see any special semantics with empty buffers. And I don't know what you are referring to about enforcement either.

From my perspective, the zero optimization stands alone and needs no other changes. OkHttp doesn't need the optimization.

**Comment by ejona86** [comment]
Oh, I see. You were talking about the optimized singleton instance for OkHttp. Yeah, it seems that would be necessary since you would be sharing it.

**Comment by louiscryan** [comment]
Failing any attempt to write to an empty buffer, even byte.length==0

If we ever implemented buffer.grow() which is not unlikely given the desire to support messages where the size is not-known a priori should we allow growth for empty buffers?

**Comment by ejona86** [comment]
Actually, instead of a special empty OkHttp buffer, we could just have OkHttpWritableBuffer.write make sure that writableBytes never goes negative. Then it would be safe to share.

We already talked about not growing Netty buffers in certain cases like if a copy would be required. I expect that our grow API will need a way to fail, so the empty buffer would just always fail.

File: core/src/test/java/io/grpc/transport/MessageFramerTest.java [hunk 2]
```diff
@@ -270,5 +270,10 @@ public int hashCode() {
     public WritableBuffer allocate(int capacityHint) {
       return new ByteWritableBuffer(Math.min(maxSize, Math.max(capacityHint, minSize)));
     }
+
+    @Override
+    public WritableBuffer empty() {
+      return new ByteWritableBuffer(0);
```
Comment (nmittler) [inline core/src/test/java/io/grpc/transport/MessageFramerTest.java]: I still think we should just make this a singleton.  Is there any reason not to?

Comment (nmittler) [inline core/src/test/java/io/grpc/transport/MessageFramerTest.java]: nm ... this is just a test :)

**Comment by nmittler** [comment]
@louiscryan LGTM ... ship it!

**Comment by louiscryan** [comment]
I've removed the empty() method from the allocator interface but I've kept the implementations.

PTAL.

**Comment by buchgr** [comment]
@louiscryan thanks. LGTM

**Comment by ejona86** [comment]
@louiscryan, LGTM

**Comment by louiscryan** [comment]
After chatting with @nmittler I  wasn't really happy with the last round so instead I've altered MessageFramer.close to return whether it wrote any data or not. The stream impls can then use that directly and we don't have to rely on waiting for an empty buffer to be written through to trigger writing trailers.

PTAL

File: core/src/main/java/io/grpc/transport/AbstractStream.java [hunk 3]
```diff
@@ -130,11 +130,13 @@ public final void flush() {
    * Closes the underlying framer.
    *
    * <p>No-op if the framer has already been closed.
+   * @return {@code true} if closing the framer caused data to be written, {@code false} otherwise.
```
Comment (nmittler) [inline core/src/main/java/io/grpc/transport/AbstractStream.java]: nit: Should we copy the docs from the framer for the `endOfStream` parameter?

File: core/src/main/java/io/grpc/transport/MessageFramer.java [hunk 4]
```diff
@@ -215,12 +215,14 @@ public boolean isClosed() {
   /**
    * Flushes and closes the framer and releases any buffers. After the framer is closed or
    * disposed, additional calls to this method will have no affect.
+   * @return true if closing the framer caused any data to be written to the sink.
    */
-  public void close() {
+  public boolean close(boolean endOfStream) {
```
Comment (nmittler) [inline core/src/main/java/io/grpc/transport/MessageFramer.java]: add docs for the `endOfStream` parameter

**Comment by nmittler** [comment]
@louiscryan LGTM .. much cleaner!

File: core/src/main/java/io/grpc/transport/AbstractStream.java (hunk 3 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/transport/AbstractStream.java]: Done

File: core/src/main/java/io/grpc/transport/MessageFramer.java (hunk 4 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/transport/MessageFramer.java]: Done

**Comment by ejona86** [comment]
NACK. I'm still looking.

**Comment by ejona86** [comment]
@louiscryan, I don't understand what the latest iteration is trying to accomplish over the previous.

These are the that I understood needed to be changed:
1) Add an else inside internalSendFrame that does release(). This fixes the real bug.
2) Optimize zero-length flushes from MessageFramer.
3) Optimize ServerImpl to avoid a flush for unary responses.

It seems like the most recent iteration removed 1 and 2, and is instead only trying to improve 3. However, 3 looks like it can be solved cleanly (< 10 lines) without any changes to MessageFramer. Then there are oddities like sendFrame() now checking for null even though null is impossible due to the `if (buffer != null)` check in `MessageFramer.commitToSync()`.

**Comment by louiscryan** [comment]
1) is now fixed by not even allocating that buffer in the first place. The use by the framer of an empty buffer to signal to the stream that it is closed but had no data to write was an ugly hack that made following the order of writes in the stream hard to follow and it is now more immediately obvious.

2) Is still there and has been moved into the commitToSink

3) Is in another PR (are you looking at the right one?)

null is not impossible now as it is called from halfClose/abort for client streams

**Comment by ejona86** [comment]
I see now 1) is fixed by the addition of a release(). I don't buy that it could be fixed by "not even allocating it in the first place." Nowhere do we state that frames with zero _readable_ bytes doesn't need to be released. And note, that is a nonsensical API because it could have non-zero _writable_ bytes (as was previously the case).

3) I actually discovered was already optimized in NettyServerStream.

This PR seems to primarily remove a call to WritableBufferAllocator.allocate(0). It seems like if you are okay with passing null, you can do that equally well from MessageFramer and avoid the allocate(0) that way. That said, I would prefer just having WritableBufferAllocator.allocate(0) and not have to realize that the provided buffer may be null.

**Comment by louiscryan** [comment]
1) No, it is not fixed by that release. That is added to defend against future changes in the framer causing us to allocate a buffer and then write 0 bytes to it (this does not happen with the current code so it can be removed). The leak was caused by AbstractServerStream receiving 0 readable-byte buffers from MessageFrame and not releasing them. This is fixed by never allowing the MessageFramer to produce a 0-readable byte buffer to the sink

Given that we no longer allow the framer to emit 0 readable byte buffers we have to unwind the brittle code used to trigger trailer writing after close. Making the signal an explicit part of the close() interface as opposed to an undocumented side-effect in MessageFramer is a definite win.

I'm not OK with passing null from to MessageFramer.sink as it is a public contract. I am OK with passing null to sendFrame as it is an internal implementation detail.

**Comment by louiscryan** [comment]
Discussed offline. We'll use null to signal closure of an empty framer to the sink, this allows us to avoid extraneous allocation or polluting the allocator interface while keeping the Sink as the primary entry point for write events into the strea

PTAL

File: core/src/main/java/io/grpc/transport/MessageFramer.java [hunk 5]
```diff
@@ -218,6 +219,10 @@ public boolean isClosed() {
    */
   public void close() {
     if (!isClosed()) {
+      if (buffer != null && buffer.readableBytes() == 0) {
```
Comment (ejona86) [inline core/src/main/java/io/grpc/transport/MessageFramer.java]: I'd just assume not have this, since it is currently impossible. If you want to keep it, put a comment that it is for future safety or some such.

File: core/src/test/java/io/grpc/transport/MessageFramerTest.java [hunk 6]
```diff
@@ -100,15 +102,17 @@ public void closeCombinedWithFullSink() {
     verifyNoMoreInteractions(sink);
     framer.close();
     verify(sink).deliverFrame(
-        toWriteBuffer(new byte[] {0, 0, 0, 0, 7, 3, 14, 1, 5, 9, 2, 6}), true, true);
+        toWriteBuffer(new byte[]{0, 0, 0, 0, 7, 3, 14, 1, 5, 9, 2, 6}), true, true);
```
Comment (ejona86) [inline core/src/test/java/io/grpc/transport/MessageFramerTest.java]: The space before the { was correct before.

**Comment by ejona86** [comment]
@louiscryan, LGTM

File: core/src/main/java/io/grpc/transport/MessageFramer.java (hunk 5 above)
Comment (nmittler) [inline core/src/main/java/io/grpc/transport/MessageFramer.java]: Do you need the `readableBytes() == 0` part of this check?

**Comment by nmittler** [comment]
LGTM

File: core/src/main/java/io/grpc/transport/MessageFramer.java (hunk 5 above)
Comment (louiscryan) [inline core/src/main/java/io/grpc/transport/MessageFramer.java]: done

File: core/src/test/java/io/grpc/transport/MessageFramerTest.java (hunk 6 above)
Comment (louiscryan) [inline core/src/test/java/io/grpc/transport/MessageFramerTest.java]: Done
